src/iosrv/sched.c

```diff
@@ -1359,10 +1384,106 @@ sched_pop_one(struct sched_data *data, ABT_pool pool, int pool_idx)
 	return unit;
 }
 
+#define SCHED_IDLE_THRESH	8000UL	/* msecs */
+
+/*
+ * Try to relax CPU for a short period when the xstream is idle. The relaxing
+ * period can't be too long, otherwise, potential external events like:
+ * incoming network requests, new ULTs created by other xstream (from the
+ * collective call or offloading call) could be delayed too much.
+ *
+ * There are also some periodical internal events from BIO, like hotplug
+ * poller, health/io stats collecting, blobstore state transition, etc. It's
+ * not easy to accurately predict the next occurrence of those events.
+ */
+static void
+sched_try_relax(struct dss_xstream *dx, ABT_pool *pools, uint32_t running)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	unsigned int		 sleep_time = sched_relax_intvl;
+	size_t			 blocked;
+	int			 ret;
+
+	dx->dx_timeout = 0;
+
+	if (info->si_stop)
+		return;
+
+	/*
+	 * There are running ULTs in current schedule cycle.
+	 *
+	 * NB. DRPC listener ULT is currently always running (it waits on
+	 * drpc_progress(), so the DRPC listener xstream will never sleep
+	 * in this function.
+	 */
+	if (running != 0)
+		return;
+
+	/* There are queued requests to be processed */
+	if (info->si_req_cnt != 0)
+		return;
+
+	ret = ABT_pool_get_total_size(pools[DSS_POOL_GENERIC], &blocked);
+	if (ret != ABT_SUCCESS) {
+		D_ERROR("XS(%d) get ABT pool(%d) total size error: %d\n",
+			dx->dx_xs_id, DSS_POOL_GENERIC, ret);
+		return;
+	}
+	D_ASSERTF(info->si_sleep_cnt + info->si_wait_cnt <= blocked,
+		  "sleep:%d + wait:%d > blocked:%zd\n",
+		  info->si_sleep_cnt, info->si_wait_cnt, blocked);
+
+	/*
+	 * Only start relaxing when all blocked ULTs are either sleeping
+	 * ULT or long wait ULT.
+	 */
+	if (blocked > info->si_sleep_cnt + info->si_wait_cnt)
+		return;
+
+	/*
+	 * System is currently idle, but we only start relaxing when there is
+	 * no external events for a short period of SCHED_IDLE_THRESH.
+	 */
+	D_ASSERT(info->si_cur_ts >= info->si_stats.ss_busy_ts);
+	if (info->si_cur_ts - info->si_stats.ss_busy_ts < SCHED_IDLE_THRESH)
+		return;
+
+	/* Adjust sleep time according to the first sleeping ULT */
+	if (info->si_sleep_cnt > 0) {
+		struct sched_request	*req;
+
+		D_ASSERT(!d_list_empty(&info->si_sleep_list));
+		req = d_list_entry(info->si_sleep_list.next,
+				   struct sched_request, sr_link);
+
+		D_ASSERT(req->sr_wakeup_time > info->si_cur_ts);
+		if (sleep_time > req->sr_wakeup_time - info->si_cur_ts)
+			sleep_time = req->sr_wakeup_time - info->si_cur_ts;
+	}
+
+	/*
+	 * Wait on external network request if the xstream has Cart context,
+	 * otherwise, sleep for a while.
+	 */
+	if (sched_relax_mode != 0 && dx->dx_comm) {
+		/* convert to micro-seconds */
+		dx->dx_timeout = sleep_time * 1000;
+	} else {
+		ret = usleep(sleep_time * 1000);
```

Comment:
[Optional] Is there any reason behind picking usleep (whose man page seems fishy)? For instance, why not clock_nanosleep?

Reply:
I'm just not sure if high-resolution sleep has more overhead, it looks to me 'micro-second' resolution is enough for us. What's your specific concern of usleep()?